lib/cosmos/interfaces/interface.rb

```diff
@@ -180,10 +180,20 @@ def write_interface
     def read
       raise "Interface not connected for read: #{@name}" unless connected? && read_allowed?
 
+      first = true
       loop do
-        # Read data for a packet
-        data = read_interface()
-        return nil unless data
+        # Protocols may have cached data for a packet, so initially just inject a blank string
+        # Otherwise we can hold off outputing other packets where all the data has already
+        # been received
+        if !first or @read_protocols.length <= 0
```

Comment:
Ok so it's more about the subsequent calls to read. The first is what got me confused. First time through the loop but not the first time reading.